eventuals/os.h

```diff
@@ -60,6 +67,171 @@ inline void SetAffinity(std::thread& thread, const size_t cpu) {
 }
 #endif // __MACH__
 
+#if defined(__linux__) || defined(__MACH__)
+
+////////////////////////////////////////////////////////////////////////
+
+namespace os {
+
+////////////////////////////////////////////////////////////////////////
+
+class Thread {
+ public:
+  Thread()
+    : thread_handle_{pthread_t{}},
+      joinable_(false) {}
+
+  ~Thread() noexcept {
+    CHECK(!joinable_) << "A thread was left not joined/not detached";
+  }
+
+  Thread(const Thread& other) = delete;
+
+  // IMPORTANT NOTE: on macos stacksize should be a multiple of the system
+  // page size!!!
+  // (check_line_length skip)
+  // https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/pthread_attr_setstacksize.3.html
+  // The buffer specified by `name` must contain a null-terminated string of
+  // 15 characters or less in length (not including the NULL). If the
+  // length of name is greater than 15 characters, the excess characters
+  // are ignored. If name is null, the thread name attribute is reset to
+  // the default.
+  // (check_line_length skip)
+  // https://www.ibm.com/docs/en/i/7.4?topic=ssw_ibm_i_74/apis/pthread_attr_setname_np.htm
+  template <typename Callable>
+  Thread(
+      Callable&& callable,
+      const std::string& name,
+      const size_t stack_size_in_bytes = 8'388'608)
+    : joinable_{true} {
+    CHECK_GE(stack_size_in_bytes, PTHREAD_STACK_MIN)
+        << "Stack size should not be less than"
+           " the system-defined minimum size";
+
+    pthread_attr_t attr;
+
+    PCHECK(pthread_attr_init(&attr) == 0)
+        << "Error in pthread_attr_init(...)";
+
+    // IMPORTANT NOTE: on macos stacksize should be a multiple of the system
+    // page size, otherwise `pthread_attr_setstacksize` will fail.
+    // (check_line_length skip)
+    // https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/pthread_attr_setstacksize.3.html
+    PCHECK(pthread_attr_setstacksize(&attr, stack_size_in_bytes) == 0)
+        << "Failed to set the stack size via 'pthread_attr_setstacksize'"
+           " (if you are on macOS - probably you are trying"
+           " to set the stack size which is not a multiple"
+           " of the system page size)";
+
+    struct Data {
+      std::string thread_name{};
+      Callable callable;
+      Data(const std::string& name, Callable&& f)
+        : thread_name{name},
+          callable{std::forward<Callable>(f)} {}
```

Comment:
Yep! This is necessary, otherwise compiler will message us out that there is no matching constructor!
![1](https://user-images.githubusercontent.com/68085133/185891709-066fde89-6a37-44cf-a316-9438de4c6d85.png)